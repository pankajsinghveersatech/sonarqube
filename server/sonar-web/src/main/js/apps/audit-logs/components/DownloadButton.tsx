/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { endOfDay, startOfDay, subDays } from 'date-fns';
import * as React from 'react';
import { translate } from '../../../helpers/l10n';
import { getBaseUrl } from '../../../helpers/system';
import '../style.css';
import { now, RangeOption } from '../utils';

export interface DownloadButtonProps {
  dateRange?: { from?: Date; to?: Date };
  downloadStarted: boolean;
  onStartDownload: () => void;
  selection: RangeOption;
}

const RANGE_OPTION_START = {
  [RangeOption.Today]: () => now(),
  [RangeOption.Week]: () => subDays(now(), 7),
  [RangeOption.Month]: () => subDays(now(), 30),
  [RangeOption.Trimester]: () => subDays(now(), 90)
};

const toISODateString = (date: Date) => date.toISOString();

function getRangeParams(selection: RangeOption, dateRange?: { from?: Date; to?: Date }) {
  if (selection === RangeOption.Custom) {
    // dateRange should be complete if 'custom' is selected
    if (!(dateRange?.to && dateRange?.from)) {
      return '';
    }

    return new URLSearchParams({
      from: toISODateString(startOfDay(dateRange.from)),
      to: toISODateString(endOfDay(dateRange.to))
    }).toString();
  }

  return new URLSearchParams({
    from: toISODateString(startOfDay(RANGE_OPTION_START[selection]())),
    to: toISODateString(now())
  }).toString();
}

export default function DownloadButton(props: DownloadButtonProps) {
  const { dateRange, downloadStarted, selection } = props;

  const downloadDisabled =
    downloadStarted ||
    (selection === RangeOption.Custom &&
      (dateRange?.from === undefined || dateRange?.to === undefined));

  const downloadUrl = downloadDisabled
    ? '#'
    : `${getBaseUrl()}/api/audit_logs/download?${getRangeParams(selection, dateRange)}`;

  return (
    <>
      <a
        className={classNames('button button-primary', { disabled: downloadDisabled })}
        download="audit_logs.json"
        onClick={downloadDisabled ? undefined : props.onStartDownload}
        href={downloadUrl}
        rel="noopener noreferrer"
        target="_blank">
        {translate('download_verb')}
      </a>

      {downloadStarted && (
        <div className="spacer-top">
          <p>{translate('audit_logs.download_start.sentence.1')}</p>
          <p>{translate('audit_logs.download_start.sentence.2')}</p>
          <br />
          <p>{translate('audit_logs.download_start.sentence.3')}</p>
        </div>
      )}
    </>
  );
}
