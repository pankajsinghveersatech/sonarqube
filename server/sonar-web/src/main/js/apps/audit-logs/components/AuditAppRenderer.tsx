/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { subDays } from 'date-fns';
import * as React from 'react';
import { Helmet } from 'react-helmet-async';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import Suggestions from '../../../app/components/embed-docs-modal/Suggestions';
import DateRangeInput from '../../../components/controls/DateRangeInput';
import Radio from '../../../components/controls/Radio';
import { translate } from '../../../helpers/l10n';
import '../style.css';
import { HousekeepingPolicy, now, RangeOption } from '../utils';
import DownloadButton from './DownloadButton';

export interface AuditAppRendererProps {
  dateRange?: { from?: Date; to?: Date };
  downloadStarted: boolean;
  handleOptionSelection: (option: RangeOption) => void;
  handleDateSelection: (dateRange: { from?: Date; to?: Date }) => void;
  handleStartDownload: () => void;
  housekeepingPolicy: HousekeepingPolicy;
  selection: RangeOption;
}

const HOUSEKEEPING_MONTH_THRESHOLD = 30;
const HOUSEKEEPING_TRIMESTER_THRESHOLD = 90;

const HOUSEKEEPING_POLICY_VALUES = {
  [HousekeepingPolicy.Weekly]: 7,
  [HousekeepingPolicy.Monthly]: 30,
  [HousekeepingPolicy.Trimestrial]: 90,
  [HousekeepingPolicy.Yearly]: 365
};

const getRangeOptions = (housekeepingPolicy: HousekeepingPolicy) => {
  const rangeOptions = [RangeOption.Today, RangeOption.Week];

  if (HOUSEKEEPING_POLICY_VALUES[housekeepingPolicy] >= HOUSEKEEPING_MONTH_THRESHOLD) {
    rangeOptions.push(RangeOption.Month);
  }

  if (HOUSEKEEPING_POLICY_VALUES[housekeepingPolicy] >= HOUSEKEEPING_TRIMESTER_THRESHOLD) {
    rangeOptions.push(RangeOption.Trimester);
  }

  rangeOptions.push(RangeOption.Custom);

  return rangeOptions;
};

export default function AuditAppRenderer(props: AuditAppRendererProps) {
  const { dateRange, downloadStarted, housekeepingPolicy, selection } = props;

  return (
    <div className="page page-limited" id="marketplace-page">
      <Suggestions suggestions="audit-logs" />
      <Helmet title={translate('audit_logs.page')} />

      <header className="page-header">
        <h1 className="page-title">{translate('audit_logs.page')}</h1>
      </header>

      <p className="big-spacer-bottom">
        {translate('audit_logs.page.description.1')}
        <br />
        <FormattedMessage
          id="audit_logs.page.description.2"
          defaultMessage={translate('audit_logs.page.description.2')}
          values={{
            housekeeping: translate('audit_logs.housekeeping_policy', housekeepingPolicy),
            link: (
              <Link
                to={{
                  pathname: '/admin/settings',
                  query: { category: 'housekeeping' },
                  hash: '#auditLogs'
                }}>
                {translate('audit_logs.page.description.link')}
              </Link>
            )
          }}
        />
      </p>

      <div className="huge-spacer-bottom">
        <h2 className="big-spacer-bottom">{translate('audit_logs.download')}</h2>

        <ul>
          {getRangeOptions(housekeepingPolicy).map(option => (
            <li key={option} className="spacer-bottom">
              <Radio
                checked={selection === option}
                onCheck={props.handleOptionSelection}
                value={option}>
                {translate('audit_logs.range_option', option)}
              </Radio>
            </li>
          ))}
        </ul>

        <DateRangeInput
          className="big-spacer-left"
          onChange={props.handleDateSelection}
          minDate={subDays(now(), HOUSEKEEPING_POLICY_VALUES[housekeepingPolicy])}
          maxDate={now()}
          value={dateRange}
        />
      </div>

      <DownloadButton
        dateRange={dateRange}
        downloadStarted={downloadStarted}
        onStartDownload={props.handleStartDownload}
        selection={selection}
      />
    </div>
  );
}
