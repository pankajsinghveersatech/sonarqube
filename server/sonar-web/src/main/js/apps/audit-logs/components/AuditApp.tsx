/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import { getAppState, getGlobalSettingValue, Store } from '../../../store/rootReducer';
import { AdminPageExtension } from '../../../types/extension';
import { fetchValues } from '../../settings/store/actions';
import '../style.css';
import { HousekeepingPolicy, RangeOption } from '../utils';
import AuditAppRenderer from './AuditAppRenderer';

interface Props {
  auditHousekeepingPolicy: HousekeepingPolicy;
  fetchValues: typeof fetchValues;
  hasGovernanceExtension?: boolean;
}

interface State {
  dateRange?: { from?: Date; to?: Date };
  downloadStarted: boolean;
  selection: RangeOption;
}

export class AuditApp extends React.PureComponent<Props, State> {
  state: State = {
    downloadStarted: false,
    selection: RangeOption.Today
  };

  componentDidMount() {
    const { hasGovernanceExtension } = this.props;
    if (hasGovernanceExtension) {
      this.props.fetchValues('sonar.dbcleaner.auditHousekeeping');
    }
  }

  handleDateSelection = (dateRange: { from?: Date; to?: Date }) =>
    this.setState({ dateRange, downloadStarted: false, selection: RangeOption.Custom });

  handleOptionSelection = (selection: RangeOption) =>
    this.setState({ dateRange: undefined, downloadStarted: false, selection });

  handleStartDownload = () => {
    setTimeout(() => {
      this.setState({ downloadStarted: true });
    }, 0);
  };

  render() {
    const { hasGovernanceExtension, auditHousekeepingPolicy } = this.props;

    return hasGovernanceExtension ? (
      <AuditAppRenderer
        handleDateSelection={this.handleDateSelection}
        handleOptionSelection={this.handleOptionSelection}
        handleStartDownload={this.handleStartDownload}
        housekeepingPolicy={auditHousekeepingPolicy || HousekeepingPolicy.Monthly}
        {...this.state}
      />
    ) : null;
  }
}

const mapDispatchToProps = { fetchValues };

const mapStateToProps = (state: Store) => {
  const settingValue = getGlobalSettingValue(state, 'sonar.dbcleaner.auditHousekeeping');
  const { adminPages } = getAppState(state);
  const hasGovernanceExtension = Boolean(
    adminPages?.find(e => e.key === AdminPageExtension.GovernanceConsole)
  );
  return {
    auditHousekeepingPolicy: settingValue?.value as HousekeepingPolicy,
    hasGovernanceExtension
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(AuditApp);
