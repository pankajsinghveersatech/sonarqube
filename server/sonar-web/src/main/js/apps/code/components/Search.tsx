/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { omit } from 'lodash';
import * as React from 'react';
import { getTree } from '../../../api/components';
import SearchBox from '../../../components/controls/SearchBox';
import { Location, Router, withRouter } from '../../../components/hoc/withRouter';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { translate } from '../../../helpers/l10n';
import { BranchLike } from '../../../types/branch-like';

interface Props {
  branchLike?: BranchLike;
  component: T.ComponentMeasure;
  location: Location;
  onSearchClear: () => void;
  onSearchResults: (results?: T.ComponentMeasure[]) => void;
  router: Router;
}

interface State {
  query: string;
  loading: boolean;
}

export class Search extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {
    query: '',
    loading: false
  };

  componentDidMount() {
    this.mounted = true;
    if (this.props.location.query.search) {
      this.handleQueryChange(this.props.location.query.search);
    }
  }

  componentDidUpdate(nextProps: Props) {
    // if the component has change, reset the current state
    if (nextProps.location.query.id !== this.props.location.query.id) {
      this.setState({
        query: '',
        loading: false
      });
      this.props.onSearchClear();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    switch (event.keyCode) {
      case 13:
      case 38:
      case 40:
        event.preventDefault();
        event.currentTarget.blur();
        break;
    }
  };

  handleSearch = (query: string) => {
    if (this.mounted) {
      const { branchLike, component, router, location } = this.props;
      this.setState({ loading: true });
      router.replace({ pathname: location.pathname, query: { ...location.query, search: query } });

      const isPortfolio = ['VW', 'SVW', 'APP'].includes(component.qualifier);
      const qualifiers = isPortfolio ? 'SVW,TRK' : 'BRC,UTS,FIL';

      getTree({
        component: component.key,
        q: query,
        s: 'qualifier,name',
        qualifiers,
        ...getBranchLikeQuery(branchLike)
      })
        .then(r => {
          if (this.mounted) {
            this.setState({
              loading: false
            });
            this.props.onSearchResults(r.components);
          }
        })
        .catch(() => {
          if (this.mounted) {
            this.setState({ loading: false });
          }
        });
    }
  };

  handleQueryChange = (query: string) => {
    const { router, location } = this.props;
    this.setState({ query });
    if (query.length === 0) {
      router.replace({ pathname: location.pathname, query: omit(location.query, 'search') });
      this.props.onSearchClear();
    } else {
      this.handleSearch(query);
    }
  };

  render() {
    const { component } = this.props;
    const { loading } = this.state;
    const isPortfolio = ['VW', 'SVW', 'APP'].includes(component.qualifier);

    return (
      <div className="code-search" id="code-search">
        <SearchBox
          minLength={3}
          onChange={this.handleQueryChange}
          onKeyDown={this.handleKeyDown}
          placeholder={translate(
            isPortfolio ? 'code.search_placeholder.portfolio' : 'code.search_placeholder'
          )}
          value={this.state.query}
        />
        <DeferredSpinner className="spacer-left" loading={loading} />
      </div>
    );
  }
}

export default withRouter(Search);
