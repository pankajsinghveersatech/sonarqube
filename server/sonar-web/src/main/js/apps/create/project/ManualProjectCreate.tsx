/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { debounce } from 'lodash';
import * as React from 'react';
import { createProject, doesComponentExists } from '../../../api/components';
import ProjectKeyInput from '../../../components/common/ProjectKeyInput';
import { SubmitButton } from '../../../components/controls/buttons';
import ValidationInput from '../../../components/controls/ValidationInput';
import { Alert } from '../../../components/ui/Alert';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import MandatoryFieldsExplanation from '../../../components/ui/MandatoryFieldsExplanation';
import { translate } from '../../../helpers/l10n';
import { PROJECT_KEY_INVALID_CHARACTERS, validateProjectKey } from '../../../helpers/projects';
import { ProjectKeyValidationResult } from '../../../types/component';
import { PROJECT_NAME_MAX_LEN } from './constants';
import CreateProjectPageHeader from './CreateProjectPageHeader';
import './ManualProjectCreate.css';

interface Props {
  branchesEnabled: boolean;
  onProjectCreate: (projectKey: string) => void;
}

interface State {
  projectName: string;
  projectNameError?: string;
  projectNameTouched?: boolean;
  projectKey: string;
  projectKeyError?: string;
  projectKeyTouched?: boolean;
  validatingProjectKey: boolean;
  submitting: boolean;
}

type ValidState = State & Required<Pick<State, 'projectKey' | 'projectName'>>;

export default class ManualProjectCreate extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      projectKey: '',
      projectName: '',
      submitting: false,
      validatingProjectKey: false
    };
    this.checkFreeKey = debounce(this.checkFreeKey, 250);
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  checkFreeKey = (key: string) => {
    this.setState({ validatingProjectKey: true });

    return doesComponentExists({ component: key })
      .then(alreadyExist => {
        if (this.mounted && key === this.state.projectKey) {
          this.setState({
            projectKeyError: alreadyExist
              ? translate('onboarding.create_project.project_key.taken')
              : undefined,
            validatingProjectKey: false
          });
        }
      })
      .catch(() => {
        if (this.mounted && key === this.state.projectKey) {
          this.setState({ projectKeyError: undefined, validatingProjectKey: false });
        }
      });
  };

  canSubmit(state: State): state is ValidState {
    const { projectKey, projectKeyError, projectName, projectNameError } = state;
    return Boolean(
      projectKeyError === undefined &&
        projectNameError === undefined &&
        projectKey.length > 0 &&
        projectName.length > 0
    );
  }

  handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const { state } = this;
    if (this.canSubmit(state)) {
      this.setState({ submitting: true });
      createProject({
        project: state.projectKey,
        name: (state.projectName || state.projectKey).trim()
      }).then(
        ({ project }) => this.props.onProjectCreate(project.key),
        () => {
          if (this.mounted) {
            this.setState({ submitting: false });
          }
        }
      );
    }
  };

  handleProjectKeyChange = (projectKey: string, fromUI = false) => {
    const projectKeyError = this.validateKey(projectKey);

    this.setState({
      projectKey,
      projectKeyError,
      projectKeyTouched: fromUI
    });

    if (projectKeyError === undefined) {
      this.checkFreeKey(projectKey);
    }
  };

  handleProjectNameChange = (projectName: string, fromUI = false) => {
    this.setState(
      {
        projectName,
        projectNameError: this.validateName(projectName),
        projectNameTouched: fromUI
      },
      () => {
        if (!this.state.projectKeyTouched) {
          const sanitizedProjectKey = this.state.projectName
            .trim()
            .replace(PROJECT_KEY_INVALID_CHARACTERS, '-');
          this.handleProjectKeyChange(sanitizedProjectKey);
        }
      }
    );
  };

  validateKey = (projectKey: string) => {
    const result = validateProjectKey(projectKey);
    return result === ProjectKeyValidationResult.Valid
      ? undefined
      : translate('onboarding.create_project.project_key.error', result);
  };

  validateName = (projectName: string) => {
    if (projectName.length === 0) {
      return translate('onboarding.create_project.display_name.error.empty');
    } else if (projectName.length > PROJECT_NAME_MAX_LEN) {
      return translate('onboarding.create_project.display_name.error.too_long');
    }
    return undefined;
  };

  render() {
    const {
      projectKey,
      projectKeyError,
      projectKeyTouched,
      projectName,
      projectNameError,
      projectNameTouched,
      validatingProjectKey,
      submitting
    } = this.state;
    const { branchesEnabled } = this.props;

    const touched = !!(projectKeyTouched || projectNameTouched);
    const projectNameIsInvalid = touched && projectNameError !== undefined;
    const projectNameIsValid = touched && projectNameError === undefined;

    return (
      <>
        <CreateProjectPageHeader title={translate('onboarding.create_project.setup_manually')} />

        <div className="create-project-manual">
          <div className="flex-1 huge-spacer-right">
            <form className="manual-project-create" onSubmit={this.handleFormSubmit}>
              <MandatoryFieldsExplanation className="big-spacer-bottom" />

              <ValidationInput
                className="form-field"
                description={translate('onboarding.create_project.display_name.description')}
                error={projectNameError}
                id="project-name"
                isInvalid={projectNameIsInvalid}
                isValid={projectNameIsValid}
                label={translate('onboarding.create_project.display_name')}
                required={true}>
                <input
                  className={classNames('input-super-large', {
                    'is-invalid': projectNameIsInvalid,
                    'is-valid': projectNameIsValid
                  })}
                  id="project-name"
                  maxLength={PROJECT_NAME_MAX_LEN}
                  minLength={1}
                  onChange={e => this.handleProjectNameChange(e.currentTarget.value, true)}
                  type="text"
                  value={projectName}
                  autoFocus={true}
                />
              </ValidationInput>
              <ProjectKeyInput
                error={projectKeyError}
                label={translate('onboarding.create_project.project_key')}
                onProjectKeyChange={e => this.handleProjectKeyChange(e.currentTarget.value, true)}
                projectKey={projectKey}
                touched={touched}
                validating={validatingProjectKey}
              />

              <SubmitButton disabled={!this.canSubmit(this.state) || submitting}>
                {translate('set_up')}
              </SubmitButton>
              <DeferredSpinner className="spacer-left" loading={submitting} />
            </form>

            {branchesEnabled && (
              <Alert variant="info" display="inline" className="big-spacer-top">
                {translate('onboarding.create_project.pr_decoration.information')}
              </Alert>
            )}
          </div>
        </div>
      </>
    );
  }
}
