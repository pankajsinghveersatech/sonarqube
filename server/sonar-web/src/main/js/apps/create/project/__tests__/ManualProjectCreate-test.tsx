/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { createProject, doesComponentExists } from '../../../../api/components';
import ProjectKeyInput from '../../../../components/common/ProjectKeyInput';
import { SubmitButton } from '../../../../components/controls/buttons';
import ValidationInput from '../../../../components/controls/ValidationInput';
import { validateProjectKey } from '../../../../helpers/projects';
import { mockEvent } from '../../../../helpers/testMocks';
import { change, submit, waitAndUpdate } from '../../../../helpers/testUtils';
import { ProjectKeyValidationResult } from '../../../../types/component';
import { PROJECT_NAME_MAX_LEN } from '../constants';
import ManualProjectCreate from '../ManualProjectCreate';

jest.mock('../../../../api/components', () => ({
  createProject: jest.fn().mockResolvedValue({ project: { key: 'bar', name: 'Bar' } }),
  doesComponentExists: jest
    .fn()
    .mockImplementation(({ component }) => Promise.resolve(component === 'exists'))
}));

jest.mock('../../../../helpers/projects', () => {
  const { PROJECT_KEY_INVALID_CHARACTERS } = jest.requireActual('../../../../helpers/projects');
  return {
    validateProjectKey: jest.fn(() => ProjectKeyValidationResult.Valid),
    PROJECT_KEY_INVALID_CHARACTERS
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot();

  const wrapper = shallowRender();
  wrapper.instance().handleProjectNameChange('My new awesome app');
  expect(wrapper).toMatchSnapshot('with form filled');

  expect(shallowRender({ branchesEnabled: true })).toMatchSnapshot('with branches enabled');
});

it('should correctly create a project', async () => {
  const onProjectCreate = jest.fn();
  const wrapper = shallowRender({ onProjectCreate });

  wrapper
    .find(ProjectKeyInput)
    .props()
    .onProjectKeyChange(mockEvent({ currentTarget: { value: 'bar' } }));
  change(wrapper.find('input#project-name'), 'Bar');
  expect(wrapper.find(SubmitButton).props().disabled).toBe(false);
  expect(validateProjectKey).toBeCalledWith('bar');
  expect(doesComponentExists).toBeCalledWith({ component: 'bar' });

  submit(wrapper.find('form'));
  expect(createProject).toBeCalledWith({
    project: 'bar',
    name: 'Bar'
  });

  await waitAndUpdate(wrapper);
  expect(onProjectCreate).toBeCalledWith('bar');
});

it('should not display any status when the name is not defined', () => {
  const wrapper = shallowRender();
  const projectNameInput = wrapper.find(ValidationInput);
  expect(projectNameInput.props().isInvalid).toBe(false);
  expect(projectNameInput.props().isValid).toBe(false);
});

it('should have an error when the key is invalid', () => {
  (validateProjectKey as jest.Mock).mockReturnValueOnce(ProjectKeyValidationResult.TooLong);

  const wrapper = shallowRender();

  wrapper.instance().handleProjectKeyChange('');
  expect(wrapper.find(ProjectKeyInput).props().error).toBe(
    `onboarding.create_project.project_key.error.${ProjectKeyValidationResult.TooLong}`
  );
});

it('should have an error when the key already exists', async () => {
  const wrapper = shallowRender();
  wrapper.instance().handleProjectKeyChange('exists', true);
  await waitAndUpdate(wrapper);
  expect(wrapper.state().projectKeyError).toBe('onboarding.create_project.project_key.taken');
});

it('should ignore promise return if value has been changed in the meantime', async () => {
  (validateProjectKey as jest.Mock)
    .mockReturnValueOnce(ProjectKeyValidationResult.Valid)
    .mockReturnValueOnce(ProjectKeyValidationResult.InvalidChar);
  const wrapper = shallowRender();
  const instance = wrapper.instance();

  instance.handleProjectKeyChange('exists', true);
  instance.handleProjectKeyChange('exists%', true);

  await waitAndUpdate(wrapper);

  expect(wrapper.state().projectKeyTouched).toBe(true);
  expect(wrapper.state().projectKeyError).toBe(
    `onboarding.create_project.project_key.error.${ProjectKeyValidationResult.InvalidChar}`
  );
});

it('should autofill the key based on the name, and sanitize it', () => {
  const wrapper = shallowRender();

  wrapper.instance().handleProjectNameChange('newName', true);
  expect(wrapper.state().projectKey).toBe('newName');

  wrapper.instance().handleProjectNameChange('my invalid +"*รง%&/()= name', true);
  expect(wrapper.state().projectKey).toBe('my-invalid-name');
});

it.each([
  ['empty', ''],
  ['too_long', new Array(PROJECT_NAME_MAX_LEN + 1).fill('a').join('')]
])('should have an error when the name is %s', (errorSuffix: string, projectName: string) => {
  const wrapper = shallowRender();

  wrapper.instance().handleProjectNameChange(projectName, true);
  expect(wrapper.find(ValidationInput).props().isInvalid).toBe(true);
  expect(wrapper.state().projectNameError).toBe(
    `onboarding.create_project.display_name.error.${errorSuffix}`
  );
});

function shallowRender(props: Partial<ManualProjectCreate['props']> = {}) {
  return shallow<ManualProjectCreate>(
    <ManualProjectCreate branchesEnabled={false} onProjectCreate={jest.fn()} {...props} />
  );
}
