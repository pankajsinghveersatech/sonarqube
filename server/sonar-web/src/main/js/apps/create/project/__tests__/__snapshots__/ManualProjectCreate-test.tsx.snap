// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Fragment>
  <CreateProjectPageHeader
    title="onboarding.create_project.setup_manually"
  />
  <div
    className="create-project-manual"
  >
    <div
      className="flex-1 huge-spacer-right"
    >
      <form
        className="manual-project-create"
        onSubmit={[Function]}
      >
        <MandatoryFieldsExplanation
          className="big-spacer-bottom"
        />
        <ValidationInput
          className="form-field"
          description="onboarding.create_project.display_name.description"
          id="project-name"
          isInvalid={false}
          isValid={false}
          label="onboarding.create_project.display_name"
          required={true}
        >
          <input
            autoFocus={true}
            className="input-super-large"
            id="project-name"
            maxLength={255}
            minLength={1}
            onChange={[Function]}
            type="text"
            value=""
          />
        </ValidationInput>
        <ProjectKeyInput
          label="onboarding.create_project.project_key"
          onProjectKeyChange={[Function]}
          projectKey=""
          touched={false}
          validating={false}
        />
        <SubmitButton
          disabled={true}
        >
          set_up
        </SubmitButton>
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </form>
    </div>
  </div>
</Fragment>
`;

exports[`should render correctly: with branches enabled 1`] = `
<Fragment>
  <CreateProjectPageHeader
    title="onboarding.create_project.setup_manually"
  />
  <div
    className="create-project-manual"
  >
    <div
      className="flex-1 huge-spacer-right"
    >
      <form
        className="manual-project-create"
        onSubmit={[Function]}
      >
        <MandatoryFieldsExplanation
          className="big-spacer-bottom"
        />
        <ValidationInput
          className="form-field"
          description="onboarding.create_project.display_name.description"
          id="project-name"
          isInvalid={false}
          isValid={false}
          label="onboarding.create_project.display_name"
          required={true}
        >
          <input
            autoFocus={true}
            className="input-super-large"
            id="project-name"
            maxLength={255}
            minLength={1}
            onChange={[Function]}
            type="text"
            value=""
          />
        </ValidationInput>
        <ProjectKeyInput
          label="onboarding.create_project.project_key"
          onProjectKeyChange={[Function]}
          projectKey=""
          touched={false}
          validating={false}
        />
        <SubmitButton
          disabled={true}
        >
          set_up
        </SubmitButton>
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </form>
      <Alert
        className="big-spacer-top"
        display="inline"
        variant="info"
      >
        onboarding.create_project.pr_decoration.information
      </Alert>
    </div>
  </div>
</Fragment>
`;

exports[`should render correctly: with form filled 1`] = `
<Fragment>
  <CreateProjectPageHeader
    title="onboarding.create_project.setup_manually"
  />
  <div
    className="create-project-manual"
  >
    <div
      className="flex-1 huge-spacer-right"
    >
      <form
        className="manual-project-create"
        onSubmit={[Function]}
      >
        <MandatoryFieldsExplanation
          className="big-spacer-bottom"
        />
        <ValidationInput
          className="form-field"
          description="onboarding.create_project.display_name.description"
          id="project-name"
          isInvalid={false}
          isValid={false}
          label="onboarding.create_project.display_name"
          required={true}
        >
          <input
            autoFocus={true}
            className="input-super-large"
            id="project-name"
            maxLength={255}
            minLength={1}
            onChange={[Function]}
            type="text"
            value="My new awesome app"
          />
        </ValidationInput>
        <ProjectKeyInput
          label="onboarding.create_project.project_key"
          onProjectKeyChange={[Function]}
          projectKey="My-new-awesome-app"
          touched={false}
          validating={true}
        />
        <SubmitButton
          disabled={false}
        >
          set_up
        </SubmitButton>
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </form>
    </div>
  </div>
</Fragment>
`;
